packages/webdriverio/src/polyfill.ts

```diff
@@ -0,0 +1,64 @@
+import logger from '@wdio/logger'
+
+const polyfillManager = new Map<WebdriverIO.Browser, PolyfillManager>()
+const log = logger('webdriverio:PolyfillManager')
+
+export function getPolyfillManager(browser: WebdriverIO.Browser) {
+    const existingPolyfillManager = polyfillManager.get(browser)
+    if (existingPolyfillManager) {
+        return existingPolyfillManager
+    }
+
+    const newContext = new PolyfillManager(browser)
+    polyfillManager.set(browser, newContext)
+    return newContext
+}
+
+/**
+ * This class is responsible for setting polyfill scripts in the browser.
+ */
+export class PolyfillManager {
+    #initialize: Promise<boolean>
+
+    constructor(browser: WebdriverIO.Browser) {
+        /**
+         * don't run setup when Bidi is not supported or running unit tests
+         */
+        if (!browser.isBidi || process.env.VITEST_WORKER_ID || browser.options?.automationProtocol !== 'webdriver') {
+            this.#initialize = Promise.resolve(true)
+            return
+        }
+
+        /**
+         * A polyfill to set `__name` to the global scope which is needed for WebdriverIO to properly
+         * execute custom (preload) scripts. When using `tsx` Esbuild runs some optimizations which
+         * assume that the file contains these global variables. This is a workaround until this issue
+         * is fixed.
+         *
+         * @see https://github.com/evanw/esbuild/issues/2605
+         */
+        const polyfill = () => {
+            const closure = new Function(
+                'var __defProp = Object.defineProperty;' +
+                'var __name = (target, value) => __defProp(target, \'name\', { value, configurable: true });' +
+                'globalThis.__name = __name;'
+            )
+            return closure()
```

Comment:
Are you sure this is all that's needed. See https://github.com/evanw/esbuild/issues/607#issuecomment-747843078:

> The `__name` function is just one of many code transforms that esbuild can make that assume functions aren't arbitrarily recompiled by converting them to a string and feeding them to `eval`. Other examples include converting `a ** b` to `__pow(a, b)` and `{...a, ...b}` to `__assign({}, a), b)`.

Reply:
Are you aware of more polyfills that are needed here?

Reply:
Ruffle doesn't need any others for its tests currently but that doesn't means others won't or that future tests won't require others. I think any fron this file might be needed:

https://github.com/evanw/esbuild/blob/main/internal/runtime/runtime.go

As stated in https://github.com/evanw/esbuild/issues/607#issuecomment-747843078
> Preserving the value of `toString()` on a function expression is not a design goal of esbuild. If that use case is important to you, it would be best to either arrange for the code to be passed as a string instead, or to use another tool that supports preserving the value of `toString()` on a function.

Reply:
Understood. The amount of usecases for WDIO vs someone compiling a TS project with Esbuild is very different, e.g. no one would export variables from the test. My idea was to add more polyfills when needed as this seems the simplest approach while keeping the codebase as simple as possible.